File: algorithms/3_hard/D/00940_distinct_subsequences_ii/solution.c
#define MOD 1000000007

int distinctSubseqII(const char *restrict s) {
    if (s == NULL) { return 0; }

    int64_t end_with[256] = { 0 };
    int64_t cnt = 0;

    for (size_t i = 0; s[i] != '\0'; i += 1) {
        const int64_t new_end_with = (cnt + 1) % MOD;
        const int64_t old_end_with = end_with[(unsigned char)s[i]];
        cnt = ((cnt - old_end_with) + new_end_with + MOD) % MOD;
        end_with[(unsigned char)s[i]] = new_end_with;
    }

    return (int)cnt;
}

File: algorithms/3_hard/D/00940_distinct_subsequences_ii/solution.rs
impl Solution {
    const MOD: i64 = 1_000_000_007;

    pub fn distinct_subseq_ii(s: String) -> i32 {
        let bytes = s.as_bytes();
        let mut end_with = [0_i64; 256];
        let mut cnt = 0_i64;

        for &b in bytes {
            let new_end_with = (cnt + 1) % Self::MOD;
            let old_end_with = end_with[b as usize];
            cnt = ((cnt - old_end_with) + new_end_with + Self::MOD) % Self::MOD;
            end_with[b as usize] = new_end_with;
        }

        cnt as i32
    }
}
